cli-plugins/manager/manager.go

```diff
@@ -0,0 +1,190 @@
+
+func (e ErrPluginNotFound) Error() string {
+	return "Error: No such CLI plugin: " + string(e)
+}
+
+var (
+	pluginDirs     []string
+	pluginDirsOnce sync.Once
+)
+
+func getPluginDirs(dockerCli command.Cli) []string {
+	pluginDirsOnce.Do(func() {
+		// Mostly for test.
+		if ds := os.Getenv("DOCKER_CLI_PLUGIN_EXTRA_DIRS"); ds != "" {
+			pluginDirs = append(pluginDirs, strings.Split(ds, ":")...)
+		}
+
+		if cfg := dockerCli.ConfigFile(); cfg != nil {
+			pluginDirs = append(pluginDirs, cfg.CLIPluginExtraDirs...)
+		}
+
+		pluginDirs = append(pluginDirs, defaultPluginDirs...)
+	})
+	return pluginDirs
+}
+
+func addPluginCandidatesFromDir(res map[string][]string, d string) error {
+	dentries, err := ioutil.ReadDir(d)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		// Portable? This seems to be as good as it gets :-/
+		if serr, ok := err.(*os.SyscallError); ok && serr.Err == syscall.ENOTDIR {
+			return nil
+		}
+		return err
+	}
+	for _, dentry := range dentries {
+		switch dentry.Mode() & os.ModeType {
+		case 0, os.ModeSymlink:
+			// Regular file or symlink, keep going
+		default:
+			// Something else, ignore.
+			continue
+		}
+		name := dentry.Name()
+		if !strings.HasPrefix(name, NamePrefix) {
+			continue
+		}
+		name = strings.TrimPrefix(name, NamePrefix)
+		if runtime.GOOS == "windows" {
+			exe := ".exe"
+			if !strings.HasSuffix(name, exe) {
+				continue
+			}
+			name = strings.TrimSuffix(name, exe)
+		}
+		res[name] = append(res[name], filepath.Join(d, dentry.Name()))
+	}
+	return nil
+}
+
+// listPluginCandidates allows the dirs to be specified for testing purposes.
+func listPluginCandidates(dirs []string) (map[string][]string, error) {
+	result := make(map[string][]string)
+	for _, d := range dirs {
+		if err := addPluginCandidatesFromDir(result, d); err != nil {
+			return nil, err // Or return partial result?
+		}
+	}
+	return result, nil
+}
+
+// ListPluginCandidates returns a map from plugin name to the list of (unvalidated) Candidates. The list is in descending order of priority.
+func ListPluginCandidates(dockerCli command.Cli) (map[string][]string, error) {
+	return listPluginCandidates(getPluginDirs(dockerCli))
+}
+
+// ListPlugins produces a list of the plugins available on the system
+func ListPlugins(dockerCli command.Cli, rootcmd *cobra.Command) ([]Plugin, error) {
+	candidates, err := ListPluginCandidates(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	var plugins []Plugin
+	for _, paths := range candidates {
+		if len(paths) == 0 {
+			continue
+		}
+		c := &candidate{paths[0]}
+		p, err := NewPlugin(c, rootcmd)
+		if err != nil {
+			return nil, err
+		}
+		p.ShadowedPaths = paths[1:]
+		plugins = append(plugins, p)
+	}
+
+	return plugins, nil
+}
+
+// FindPlugin finds a valid plugin, if the first candidate is invalid then returns an error
+func FindPlugin(dockerCli command.Cli, name string, rootcmd *cobra.Command, includeShadowed bool) (Plugin, error) {
```

Comment:
Should this function be exported? It's only used in PluginRunCommand.

Reply:
I think it was used in `docker cli-plugins inspect`, but since I've now dropped that you are correct.